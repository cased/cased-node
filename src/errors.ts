export class AuthenticationError extends Error {
  constructor(message: string) {
    super(message);
    Object.setPrototypeOf(this, AuthenticationError.prototype);
  }
}

interface ErrorResponse {
  error: string;
  message?: string;
  messages?: { [name: string]: string[] };
}

export class APIError extends Error {
  constructor(response: ErrorResponse) {
    let message = "unknown API error";

    if (response.message) {
      message = response.message;
    } else if (response.messages) {
      let messages: string[] = [];
      for (const name in response.messages) {
        messages.push(`${name} ${response.messages[name]}`);
      }
      message = messages.join(", ");
    }

    super(message);
    Object.setPrototypeOf(this, APIError.prototype);
  }
}

export class ExportError extends Error {
  constructor(message: string) {
    super(message);
    Object.setPrototypeOf(this, ExportError.prototype);
  }
}

export class ReasonRequiredError extends Error {
  constructor() {
    super("Reason required");
    Object.setPrototypeOf(this, ReasonRequiredError.prototype);
  }
}
