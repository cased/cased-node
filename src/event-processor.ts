import type { AuditEvent } from "./event";
import { randomBytes } from "crypto";
import { EventProcessor } from ".";
import type { IncomingMessage } from "http";

/**
 * Function that modifies or add keys to published events.
 *
 * @public
 *
 * ```ts
 * const myProcessor = (event) => {
 *   return { ...event, foo: 42 }
 * }
 * ```
 */
export type EventProcessor = (event: Readonly<AuditEvent>) => AuditEvent;

/**
 * Combine a pipeline of event processors into a single processor.
 *
 * @public
 * @param eventProcessors - A list of event processors.
 * @returns A single event processor.
 */
export const composeEventProcessor = (
  ...eventProcessors: EventProcessor[]
): EventProcessor => {
  return (event) => {
    for (const eventProcessor of eventProcessors) {
      event = eventProcessor(event);
    }
    return event;
  };
};

/**
 * Add Cased conventional keys to every published event.
 *
 * @public
 * @param event - An event.
 * @returns An event adding a `case_id` and `timestamp` property.
 */
export const defaultEventProcessor: EventProcessor = (event) => {
  return {
    cased_id: randomBytes(16).toString("hex"),
    timestamp: new Date().toISOString(),
    ...event,
  };
};

/**
 * Get Cased conventional keys for an HTTP Request.
 *
 * @param req - An HTTP incoming request.
 * @returns An event.
 */
export const requestEvent = (req: IncomingMessage): AuditEvent => {
  return {
    location: req.connection.remoteAddress,
    http_user_agent: req.headers["user-agent"],
    http_url: req.url,
    http_method: req.method,
  };
};

/**
 * Mask sensitive fields of an audit event.
 *
 * @public
 * @param field - An object key to mask.
 * @returns An event with PII masked.
 *
 * ```ts
 * sensitiveField("password")
 * ```
 */
export const sensitiveField = (field: string): EventProcessor => {
  return (event) => {
    let pii: PII = {};

    const traverse = (event: AuditEvent) => {
      for (const key in event) {
        const value = event[key];

        if (key === ".cased") {
          continue;
        } else if (typeof value === "string" && key == field) {
          pii[key] = [
            {
              label: field,
              begin: 0,
              end: value.length,
            },
          ];
        } else if (typeof value === "object") {
          traverse(value);
        }
      }
    };

    traverse(event);
    return appendPII(event, pii);
  };
};

/**
 * Mask sensitive fields of an audit event that match a pattern.
 *
 * @public
 * @param label - A label for masked PII.
 * @param pattern - A pattern to match against values.
 * @returns An event with PII masked.
 *
 * ```ts
 * sensitivePattern("username", /@([A-Za-z0-9_]+)/g)
 * ```
 */
export const sensitivePattern = (
  label: string,
  pattern: RegExp
): EventProcessor => {
  return (event) => {
    let pii: PII = {};

    const traverse = (event: AuditEvent) => {
      for (const key in event) {
        const value = event[key];

        if (key === ".cased") {
          continue;
        } else if (typeof value === "string") {
          const piiRanges = matches(value, pattern).map((match) => {
            return {
              label,
              begin: match.index,
              end: match.index + match[0].length,
            };
          });
          if (piiRanges.length > 0) pii[key] = piiRanges;
        } else if (typeof value === "object") {
          traverse(value);
        }
      }
    };

    traverse(event);
    return appendPII(event, pii);
  };
};

/**
 * Structure for defining the range and label of a PII mask.
 *
 * @private
 */
interface PIIRange {
  begin: number;
  end: number;
  label: string;
}

/**
 * A set of PII keys and masking ranges.
 *
 * @private
 */
interface PII {
  [key: string]: PIIRange[] | undefined;
}

/**
 * Append additional PII ranges to event.
 *
 * @private
 * @param event - An Event
 * @param pii - PII Configuration
 */
const appendPII = (event: any, pii: PII): AuditEvent => {
  const anyEvent: any = {
    ...event,
    ".cased": { ...event[".cased"], pii: { ...event[".cased"]?.pii, ...pii } },
  };
  return anyEvent;
};

/**
 * Return all match objects for a global RegExp.
 *
 * @private
 * @param value - String to match
 * @param regexp - A regexp to test string against.
 */
const matches = (value: string, regexp: RegExp): RegExpExecArray[] => {
  let matches: RegExpExecArray[] = [];
  let match: RegExpExecArray | null;
  while ((match = regexp.exec(value)) !== null) matches.push(match);
  return matches;
};
