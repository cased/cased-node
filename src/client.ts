import * as querystring from "querystring";
import assert from "assert";
import { httpsRequest } from "./request";
import {
  config as globalConfig,
  getPolicyKey,
  Config,
  ConfigOverrides,
} from "./config";
import type { OutgoingHttpHeaders } from "http";
import { APIError, AuthenticationError } from "./errors";
import version from "./version";

/**
 * Options for making a Cased API request.
 *
 * @protected
 */
interface RequestOptions {
  /**
   * The HTTP method.
   */
  method: "GET" | "POST" | "PUT" | "DELETE";

  /**
   * Base URL for request.
   *
   * Default `"api"`.
   */
  baseURL?: "api" | "publish" | "cli";

  /**
   * The API endpoint path.
   *
   * Omit a leading and trailing slash.
   *
   * @example `"events"`, `"events/123"`
   */
  path?: string;

  /**
   * URL search query or POST data parameters.
   */
  params: any;

  /**
   * Configuration to override on global config.
   */
  config: ConfigOverrides;
}

/**
 * Make a Cased API request.
 *
 * A high level request helper that covers all API access.
 *
 * @protected
 */
export const request = async <T>(options: RequestOptions): Promise<T> => {
  const config: Config = {
    ...globalConfig,
    ...options.config,
  };

  let apiKey: string;
  if (options.baseURL === "publish") {
    if (!config.publishKey) {
      throw new AuthenticationError(
        "No Publish API key provided. " +
          "Set your API key using config.publishKey = <API-KEY>, or supply the key with the request. " +
          "See https://docs.cased.com for more information."
      );
    }
    apiKey = config.publishKey;
  } else if (options.baseURL === "cli") {
    if (!config.guardApplicationKey) {
      throw new AuthenticationError(
        "No CLI API key provided. " +
          "Set your API key using config.guardApplicationKey = <API-KEY>, or supply the key with the request. " +
          "See https://docs.cased.com for more information."
      );
    }
    apiKey = config.guardApplicationKey;

    if (!options.params.user_token) {
      options.params.user_token = options.params.userToken || config.guardUserToken;
    }
  } else {
    let policyKey = getPolicyKey(config);
    if (!policyKey) {
      throw new AuthenticationError(
        "No API key provided. " +
          "Set your API key using config.policyKey = <POLICY-KEY>, or supply the key with the request. " +
          "See https://docs.cased.com for more information."
      );
    }
    apiKey = policyKey;
  }

  const baseURL: URL =
    options.baseURL === "publish" ? config.publishBaseURL : config.apiBaseURL;

  let url = new URL(options.path || "", baseURL);

  if (options.method === "GET") {
    url.search = querystring.stringify(options.params);
  }

  let headers: OutgoingHttpHeaders = {};
  headers["Accept"] = "application/json";
  headers["Authorization"] = `Bearer ${apiKey}`;
  headers["User-Agent"] = `cased-node/${version}`;

  let postData: Buffer | null = null;

  if (options.method !== "GET") {
    postData = Buffer.from(JSON.stringify(options.params));
    headers["Content-Type"] = "application/json; charset=utf-8";
    headers["Content-Length"] = `${postData.length}`;
  }

  const opts = {
    host: baseURL.hostname,
    path: `${url.pathname}${url.search}`,
    method: options.method,
    headers: headers,
  };

  const { response, body } = await httpsRequest(opts, postData);

  let data: any | null = null;
  if (response.headers["content-type"] === "application/json; charset=utf-8") {
    data = JSON.parse(body.toString("utf8"));
  }

  switch (response.statusCode) {
    case 200:
    case 201:
    case 202:
    case 204:
    case 400:
    case 404:
      return data;

    case 302:
      data = { location: response.headers["location"] };
      return data;

    case 401:
    case 417:
      throw new APIError(data);

    default:
      assert.fail(`unhandled status code: ${response.statusCode}`);
  }
};

// Pagination

/**
 * Raw response structure of all paginated responses.
 *
 * This structure is not exposed publically. See `Page` interface instead.
 *
 * @private
 */
interface PageResponse<T> {
  results: T[];
  total_count: number;
  total_pages: number;
}

/**
 * Generic paginated results value.
 *
 * @public
 *
 * @example
 * ```ts
 * const page = await event.search({ action: "team.update" });
 * page.totalCount // 42
 *
 * // log each event for this page of results
 * for (const event of page) {
 *    console.log(event)
 * }
 *
 * // fetch second page of results
 * const secondPage = await page.nextPage();
 * ```
 */
export interface Page<T> {
  /**
   * Iterator for page items.
   *
   * @public
   *
   * @example
   * ```ts
   * const page = await event.search({ action: "team.update" });
   *
   * // log each event for this page of results
   * for (const event of page) {
   *    console.log(event)
   * }
   * ```
   */
  [Symbol.iterator](): Iterator<T, void, undefined>;

  /**
   * Async iterator for all items.
   *
   * @public
   *
   * @example
   * ```ts
   * const events = await event.search({ action: "team.update" });
   *
   * // log each event while fetching all pages
   * for async (const event of events) {
   *    console.log(event)
   * }
   * ```
   */
  [Symbol.asyncIterator](): AsyncIterator<T, void, undefined>;

  /**
   * Total number of results in data set.
   *
   * @public
   *
   * @example
   * ```ts
   * const results = await event.search({ action: "team.update" });
   * results.totalCount // 42
   * ```
   */
  totalCount: number;

  /**
   * Total number of pages in data set.
   *
   * @public
   *
   * @example
   * ```ts
   * const results = await event.search({ action: "team.update" });
   * results.pageCount // 42
   * ```
   */
  pageCount: number;

  /**
   * Fetch next page of results.
   *
   * Returns null if there are no more pages.
   *
   * @public
   *
   * @example
   * ```ts
   * const page = await event.search({ action: "team.update" });
   *
   * // fetch second page of results
   * const secondPage = await page.nextPage();
   * ```
   */
  nextPage(): Promise<Page<T> | null>;

  /**
   * Fetch previous page of results.
   *
   * Returns null if there are no previous pages.
   *
   * @public
   *
   * @example
   * ```ts
   * const secondPage = await event.search({ action: "team.update" }, { page: 2 });
   *
   * // go back to page 1
   * const firstPage = await page.previousPage();
   * ```
   */
  previousPage(): Promise<Page<T> | null>;

  /**
   * Fetch first page of results.
   *
   * @public
   *
   * @example
   * ```ts
   * const secondPage = await event.search({ action: "team.update" }, { page: 2 });
   *
   * // go back to page 1
   * const firstPage = await page.firstPage();
   * ```
   */
  firstPage(): Promise<Page<T>>;

  /**
   * Fetch last page of results.
   *
   * @public
   *
   * @example
   * ```ts
   * const firstPage = await event.search({ action: "team.update" });
   *
   * // jump to last page
   * const lastPage = await page.lastPage();
   * ```
   */
  lastPage(): Promise<Page<T>>;
}

/**
 * Standard options for any paginated request.
 *
 * @public
 */
export interface PaginationOptions {
  /**
   * Current page number.
   *
   * Default: 1
   *
   * @example
   * ```ts
   * const secondPage = await event.search({ action: "team.update" }, { page: 2 });
   * ```
   */
  page?: number;

  /**
   * Number of results per page.
   *
   * Default: 25
   *
   * @example
   * ```ts
   * const smallPage = await event.search({ action: "team.update" }, { page: 5 });
   * ```
   */
  perPage?: number;
}

/**
 * Make a paginated Cased API request.
 *
 * @protected
 */
export const requestPage = async <T>(
  options: RequestOptions & PaginationOptions
): Promise<Page<T>> => {
  const page = options.page || 1;
  const perPage = options.perPage || 25;

  const response = await request<PageResponse<T>>({
    ...options,
    params: {
      page,
      per_page: perPage,
      ...options.params,
    },
  });

  assert(response.total_count >= 0);
  assert(response.total_pages > 0);
  assert(response.results.length >= 0);

  const iterator = (): Iterator<T> => response.results[Symbol.iterator]();

  async function* asyncIterator() {
    for (let p = page; p < response.total_pages; p++) {
      for (const item of await requestPage<T>({ ...options, page: p })) {
        yield item;
      }
    }
    return;
  }

  return {
    [Symbol.iterator]: iterator,
    [Symbol.asyncIterator]: asyncIterator,
    totalCount: response.total_count,
    pageCount: response.total_pages,
    nextPage() {
      if (page >= response.total_pages) return Promise.resolve(null);
      return requestPage({ ...options, page: page + 1 });
    },
    previousPage() {
      if (page <= 1) return Promise.resolve(null);
      return requestPage({ ...options, page: page - 1 });
    },
    firstPage() {
      return requestPage({ ...options, page: 1 });
    },
    lastPage() {
      return requestPage({
        ...options,
        page: response.total_pages,
      });
    },
  };
};
