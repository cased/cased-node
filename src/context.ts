import type { AuditEvent } from "./event";
import type { Handler } from "express";
import { EventProcessor, requestEvent } from "./event-processor";
import assert from "assert";

// Define no-op functions when async_hooks aren't available
export let eventProcessor: EventProcessor = (event) => {
  return event;
};
export let middleware: Handler = (req, res, next) => {
  next();
};

export let push = (event: AuditEvent) => {};
export let pop = () => {};

try {
  const { AsyncLocalStorage } = require("async_hooks");
  const storage = new AsyncLocalStorage();
  const requestStorage = new AsyncLocalStorage();

  push = (event: AuditEvent) => {
    let events: AuditEvent[] | undefined = storage.getStore();
    if (!events) {
      events = [];
      storage.enterWith(events);
    }
    events.push(event);
  };

  pop = () => {
    const events: AuditEvent[] | undefined = storage.getStore();
    assert(events);
    events.pop();
  };

  eventProcessor = (event: AuditEvent): AuditEvent => {
    const events: AuditEvent[] | undefined = storage.getStore();
    if (!events) return event;

    let mergedEvent = event;
    for (const contextEvent of events) {
      mergedEvent = { ...contextEvent, ...mergedEvent };
    }

    return mergedEvent;
  };

  middleware = (req, res, next) => {
    push(requestEvent(req));
    next();
  };
} catch (error) {}
