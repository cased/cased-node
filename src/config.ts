import type { AuditEvent } from "./event";
import { EventProcessor, defaultEventProcessor } from "./event-processor";
import { eventProcessor as contextEventProcessor } from "./context";
import { CLI } from ".";

/**
 * Cased client configuration.
 *
 * @public
 */
export interface Config {
  /**
   * A policy name for this config.
   */
  name: string | null;

  /**
   * The API endpoint.
   *
   * "https://api.cased.com/"
   */
  apiBaseURL: URL;

  /**
   * The publish API endpoint.
   *
   * "https://publish.cased.com/"
   */
  publishBaseURL: URL;

  /**
   * The API version number.
   */
  apiVersion: number;

  /**
   * Your default policy API Key.
   */
  policyKey: string | null;

  /**
   * Named policy API Keys.
   */
  policyKeys: {
    [key: string]: string | undefined;
  };

  /**
   * Your publishing API Key.
   */
  publishKey: string | null;

  /**
   * A list of processor functions to apply to published events.
   */
  eventPipeline: EventProcessor[];

  /**
   * Mock out any publish event during testing.
   */
  mockPublishEvents: MockPublishEvents | boolean;

  /**
   * Your CLI API Key.
   */
  guardApplicationKey: string | null;

  /**
   * Your user identifier.
   */
  guardUserToken: string | null;

  guardDenyIfUnreachable: boolean;

  cli: CLI;
}

export interface CLI {
  metadata: Dictionary;
}

/**
 * A simple dictionary conforms to this interface.
 *
 * @public
 */
export interface Dictionary {
  [name: string]: string | undefined;
}

/**
 * A mocked events object.
 *
 * A simple Array conforms to this interface.
 *
 * @public
 */
export interface MockPublishEvents {
  push(event: AuditEvent): void;
}

/**
 * Configure options to override.
 *
 * @public
 */
export interface ConfigOverrides {
  /**
   * A policy name for this config.
   */
  name?: string;

  /**
   * The API endpoint.
   *
   * "https://api.cased.com/"
   */
  apiBaseURL?: URL;

  /**
   * The publish API endpoint.
   *
   * "https://publish.cased.com/"
   */
  publishBaseURL?: URL;

  /**
   * The API version number.
   */
  apiVersion?: number;

  /**
   * Your default policy API Key.
   */
  policyKey?: string;

  /**
   * Your publishing API Key.
   */
  publishKey?: string;

  /**
   * A list of processor functions to apply to published events.
   */
  eventPipeline?: EventProcessor[];

  /**
   * Mock out any publish event during testing.
   */
  mockPublishEvents?: MockPublishEvents | boolean;

  guardApplicationKey?: string;

  guardUserToken?: string;
}

const castBool = (val: string | undefined): boolean => {
  if (!val) {
    return false;
  }

  return ["true", "1"].indexOf(val) >= 0;
}

/**
 * Global configuration for process.
 *
 * A mutable binding is exported to allow for global config changes.
 *
 * ```ts
 * import { config } from "cased";
 * config.apiKey = "cs_live_f764a5f252aaca986b0526b42a6f7e95";
 * ```
 *
 * @public
 */
export let config: Config = {
  name: null,
  apiBaseURL: new URL(process.env["CASED_API_URL"] || "https://api.cased.com/"),
  publishBaseURL: new URL(
    process.env["CASED_PUBLISH_URL"] || "https://publish.cased.com/"
  ),
  apiVersion: 1,
  policyKey: process.env["CASED_POLICY_KEY"] || null,
  policyKeys: {},
  publishKey: process.env["CASED_PUBLISH_KEY"] || null,
  eventPipeline: [defaultEventProcessor, contextEventProcessor],
  mockPublishEvents: false,
  guardApplicationKey: process.env["GUARD_APPLICATION_KEY"] || null,
  guardUserToken: process.env["GUARD_USER_TOKEN"] || null,
  guardDenyIfUnreachable: castBool(process.env["DENY_IF_UNREACHABLE"]),
  cli: {
    metadata: {},
  },
};

/**
 * Get policy key for config.
 *
 * Try to read policy key from config, but fallback to fetching
 * from the environment.
 *
 * @protected
 * @param config - A policy config.
 */
export const getPolicyKey = (config: Config): string | null => {
  let key: string | undefined | null;

  if (config.name) {
    key = config.policyKeys[config.name];
    if (key) return key;

    key = process.env[`CASED_${config.name}_POLICY_KEY`];
    if (key) return key;
  }

  key = config.policyKey;
  if (key !== null) return key;

  key = process.env["CASED_POLICY_KEY"];
  if (key) return key;

  return null;
};
