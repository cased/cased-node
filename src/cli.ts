import { config as globalConfig, ConfigOverrides, Dictionary } from "./config";
import { request } from "./client";


/**
 * An `Export`s metadata and current state
 *
 * @public
 */
 export interface SessionResult {
  // The CLI session ID
  id: string;

  // The CLI session web URL
  url: string;

  // The CLI session API URL
  api_url: string;

  // The CLI session record API URL
  api_record_url: string;

  // The current state the CLI session is in
  state: "requested" | "approved" | "denied" | "timed_out" | "canceled";

  // Command that invoked CLI session.
  command: string;

  // Additional user supplied metadata about the CLI session.
  metadata: string;
  reason: string;

  // The forwarded IP V4 or IP V6 address of the user that initiated the
  // CLI session.
  forwarded_ip_address: string | null;

  // The client's IP V4 or IP V6 address that initiated the CLI session.
  ip_address: string | null;

  // The Cased user that requested the CLI session.
  requester: string;

  // The Cased user that requested the CLI session.
  responded_at: string;

  // The Cased user that responded to the CLI session.
  responder: string;

  // The CLI application that the CLI session belongs to.
  guard_application: string;

  updated_at: string;
  created_at: string;
}

/**
 * Options for creating a new `Session`.
 *
 * @public
 */
export interface SessionOptions {
  metadata?: Dictionary;
  command?: string;
  reason?: string;
}

export const session = (
  options: SessionOptions = {},
  config: ConfigOverrides = {},
): Promise<SessionResult> => {
  return request<SessionResult>({
    method: "POST",
    baseURL: "api",
    path: "cli/sessions",
    params: options,
    config: config,
  });
};
