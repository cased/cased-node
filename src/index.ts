import { config } from "./config";
export * as Context from "./context";
export * as Event from "./event";
export * as EventProcessor from "./event-processor";
export * as Export from "./export";
export * as CLI from "./cli";

export { config };
