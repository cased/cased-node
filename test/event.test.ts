jest.mock("../src/request");
const { httpsRequest } = require("../src/request");

import { Event } from "../src/index";
import { mockResponse } from "./mocks";
import assert from "assert";

test("publish", async () => {
  httpsRequest.mockImplementationOnce(
    mockResponse(200, {
      recieved_at: "1590801840.9284723",
      id: "a0e10024-7b43-43a9-bb74-e924773e876d",
      queued: {
        event_id: "a0e10024-7b43-43a9-bb74-e924773e876d",
        audit_event: { action: "user.login" },
        api_key: "cs_test_001",
      },
      status_url:
        "http://cased.localhost/api/events/a0e10024-7b43-43a9-bb74-e924773e876d",
    })
  );

  const response = await Event.publish({ action: "user.login" });
  expect(response).not.toBeNull();
});

test("publish mocked with helper", async () => {
  let events: Event.AuditEvent[] = [];
  expect(events.length).toBe(0);

  const response = await Event.publish(
    { action: "user.login" },
    { mockPublishEvents: events }
  );
  expect(response).not.toBeNull();
  expect(events.length).toBe(1);
});

describe("search", () => {
  test("raw phrase", async () => {
    httpsRequest.mockImplementationOnce(
      mockResponse(200, {
        total_count: 30,
        total_pages: 2,
        results: Array(25),
      })
    );

    const firstPage = await Event.search("action:team.update");
    expect(firstPage.totalCount).toEqual(30);
    expect(firstPage.pageCount).toEqual(2);

    {
      let eventsCount = 0;
      for (let e of firstPage) eventsCount++;
      expect(eventsCount).toEqual(25);
    }

    httpsRequest.mockImplementationOnce(
      mockResponse(200, {
        total_count: 30,
        total_pages: 2,
        results: Array(5),
      })
    );

    const secondPage = await firstPage.nextPage();
    assert(secondPage);
    expect(secondPage).toBeTruthy();
    expect(secondPage.totalCount).toEqual(30);
    expect(secondPage.pageCount).toEqual(2);

    {
      let eventsCount = 0;
      for (let e of secondPage) eventsCount++;
      expect(eventsCount).toEqual(5);
    }

    expect(secondPage.nextPage()).resolves.toBeNull();
  });

  test("phrase object", async () => {
    httpsRequest.mockImplementationOnce(
      mockResponse(200, {
        total_count: 352,
        total_pages: 15,
        results: Array(25),
      })
    );

    const page = await Event.search({ action: "team.update" });

    expect(page.totalCount).toEqual(352);
    expect(page.pageCount).toEqual(15);

    let eventsCount = 0;
    for (let e of page) eventsCount++;
    expect(eventsCount).toEqual(25);
  });

  test("action", async () => {
    httpsRequest.mockImplementationOnce(
      mockResponse(200, {
        total_count: 100,
        total_pages: 4,
        results: Array(25),
      })
    );

    const page = await Event.searchAction("user.login");

    expect(page.totalCount).toEqual(100);
    expect(page.pageCount).toEqual(4);

    let eventsCount = 0;
    for (let e of page) eventsCount++;
    expect(eventsCount).toEqual(25);
  });

  test("actor", async () => {
    httpsRequest.mockImplementationOnce(
      mockResponse(200, {
        total_count: 100,
        total_pages: 4,
        results: Array(25),
      })
    );

    const page = await Event.searchActor("jill");

    expect(page.totalCount).toEqual(100);
    expect(page.pageCount).toEqual(4);

    let eventsCount = 0;
    for (let e of page) eventsCount++;
    expect(eventsCount).toEqual(25);
  });

  const usersPolicy = {
    name: "users",
    policyKey: "cs_test_003",
  };

  test("for policy name", async () => {
    httpsRequest.mockImplementationOnce(
      mockResponse(200, {
        total_count: 352,
        total_pages: 15,
        results: Array(25),
      })
    );

    const page = await Event.search(
      { action: "team.update" },
      {},
      { name: "organizations" }
    );

    expect(page.totalCount).toEqual(352);
    expect(page.pageCount).toEqual(15);

    let eventsCount = 0;
    for (let e of page) eventsCount++;
    expect(eventsCount).toEqual(25);
  });

  test("for policy object", async () => {
    httpsRequest.mockImplementationOnce(
      mockResponse(200, {
        total_count: 352,
        total_pages: 15,
        results: Array(25),
      })
    );

    const page = await Event.search({ action: "team.update" }, {}, usersPolicy);

    expect(page.totalCount).toEqual(352);
    expect(page.pageCount).toEqual(15);

    let eventsCount = 0;
    for (let e of page) eventsCount++;
    expect(eventsCount).toEqual(25);
  });
});

test("fetch", async () => {
  httpsRequest.mockImplementationOnce(
    mockResponse(200, {
      audit_trail: {
        id: "00000000-4086-425b-83f3-d6defafb7381",
        name: "system",
      },
      audit_event_url:
        "http://cased.localhost/api/audit-trails/system/events/00000b3aa6d6672171da2f209df91831",
      audit_event: {
        action: "audit_trail_export.success",
        audit_trail_export: "",
        audit_trail_export_id:
          "AuditTrailExport;00000000-f79b-4c21-a75f-d19f35266b92",
        cased_id: "00000b3aa6d6672171da2f209df91831",
        timestamp: "2020-06-08T00:00:00Z",
      },
    })
  );

  const response = await Event.fetch(
    "system",
    "00000b3aa6d6672171da2f209df91831"
  );

  expect(response.audit_trail.name).toBe("system");
  expect(response.audit_trail.id).toBe("00000000-4086-425b-83f3-d6defafb7381");
});
