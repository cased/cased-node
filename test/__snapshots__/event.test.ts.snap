// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch 1`] = `
Object {
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer cs_test_001",
    "User-Agent": "Jest/1.0",
  },
  "host": "cased.localhost",
  "method": "GET",
  "path": "/api/audit-trails/system/events/00000b3aa6d6672171da2f209df91831",
}
`;

exports[`publish 1`] = `
Object {
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer publish_test_001",
    "Content-Length": "108",
    "Content-Type": "application/json; charset=utf-8",
    "User-Agent": "Jest/1.0",
  },
  "host": "cased.localhost",
  "method": "POST",
  "path": "/api/",
}
`;

exports[`publish 2`] = `"{\\"cased_id\\":\\"11111111111111111111111111111111\\",\\"timestamp\\":\\"2020-01-01T12:00:00.000Z\\",\\"action\\":\\"user.login\\"}"`;

exports[`search action 1`] = `
Object {
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer cs_test_001",
    "User-Agent": "Jest/1.0",
  },
  "host": "cased.localhost",
  "method": "GET",
  "path": "/api/events?page=1&per_page=25&phrase=action%3Auser.login",
}
`;

exports[`search actor 1`] = `
Object {
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer cs_test_001",
    "User-Agent": "Jest/1.0",
  },
  "host": "cased.localhost",
  "method": "GET",
  "path": "/api/events?page=1&per_page=25&phrase=actor%3Ajill",
}
`;

exports[`search for policy name 1`] = `
Object {
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer cs_test_002",
    "User-Agent": "Jest/1.0",
  },
  "host": "cased.localhost",
  "method": "GET",
  "path": "/api/events?page=1&per_page=25&phrase=action%3Ateam.update",
}
`;

exports[`search for policy object 1`] = `
Object {
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer cs_test_003",
    "User-Agent": "Jest/1.0",
  },
  "host": "cased.localhost",
  "method": "GET",
  "path": "/api/events?page=1&per_page=25&phrase=action%3Ateam.update",
}
`;

exports[`search phrase object 1`] = `
Object {
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer cs_test_001",
    "User-Agent": "Jest/1.0",
  },
  "host": "cased.localhost",
  "method": "GET",
  "path": "/api/events?page=1&per_page=25&phrase=action%3Ateam.update",
}
`;

exports[`search raw phrase 1`] = `
Object {
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer cs_test_001",
    "User-Agent": "Jest/1.0",
  },
  "host": "cased.localhost",
  "method": "GET",
  "path": "/api/events?page=1&per_page=25&phrase=action%3Ateam.update",
}
`;

exports[`search raw phrase 2`] = `
Object {
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer cs_test_001",
    "User-Agent": "Jest/1.0",
  },
  "host": "cased.localhost",
  "method": "GET",
  "path": "/api/events?page=2&per_page=25&phrase=action%3Ateam.update",
}
`;
